src/loadsave.jl

```diff
@@ -1,37 +1,119 @@
+    path = Base.find_in_path(string(pkg)) # hacky way to determine if a Package is installed
+    path == nothing && return false
+    return isfile(path)
+end
+
+function save_import(pkg::Symbol)
+    !is_installed(pkg)      && throw(NotInstalledError(pkg, ""))
+    !isdefined(Main, pkg)   && eval(Main, Expr(:import, pkg))
+    return Main.(pkg)
+end
+
+
+for (applicable_, add_, dict_) in (
+        (:applicable_loaders, :add_loader, :sym2loader),
+        (:applicable_savers,  :add_saver,  :sym2saver))
     @eval begin
-        $appl{sym}(::Formatted{DataFormat{sym}}) = get($dct, sym, [:FileIO]) # if no loader is declared, fallback to FileIO
-        function $fchk(pkg::Symbol)
-            !isdefined(Main, pkg) && eval(Main, Expr(:import, pkg))
-            return Main.(pkg)
-        end
-        function $fadd{sym}(::Type{DataFormat{sym}}, pkg::Symbol)
-            list = get($dct, sym, Symbol[])
-            $dct[sym] = push!(list, pkg)
+        $applicable_{sym}(::@compat(Union{Type{DataFormat{sym}}, Formatted{DataFormat{sym}}})) = get($dict_, sym, [:FileIO]) # if no loader is declared, fallback to FileIO
+        function $add_{sym}(::Type{DataFormat{sym}}, pkg::Symbol)
+            list = get($dict_, sym, Symbol[])
+            $dict_[sym] = push!(list, pkg)
         end
     end
 end
 
-applicable_mime{sym}(::MIME{sym}) = get(mimedict, sym, [:nothing])
-function check_mime(pkg::Symbol)
-    pkg == :nothing && error("No MIME package available")
-    !isdefined(Main, pkg) && eval(Main, Expr(:import, pkg))
-    return pkg
-end
+applicable_mime{sym}(::MIME{sym}) = get(mimedict, sym, [(Any, :FileIO)])
 
 @doc """
 `add_mime(mime, T, :Package)`  triggers `using Package` before attempting to write object of type `T` in format `mime`.
 """ ->
 function add_mime{sym,T}(::MIME{sym}, ::Type{T}, pkg::Symbol)
-    list = get(mimedict, sym, Any[])
+    list = get(mimedict, sym, @compat(Tuple{DataType, Symbol})[])
     mimedict[sym] = push!(list, (T,pkg))
 end
 
 @doc "`add_loader(fmt, :Package)` triggers `using Package` before loading format `fmt`" -> add_loader
 @doc "`add_saver(fmt, :Package)` triggers `using Package` before saving format `fmt`" -> add_saver
+
+
+@doc """
+- `load(filename)` loads the contents of a formatted file, trying to infer
+the format from `filename` and/or magic bytes in the file.
+- `load(strm)` loads from an `IOStream` or similar object. In this case,
+the magic bytes are essential.
+- `load(File(format"PNG",filename))` specifies the format directly, and bypasses inference.
+- `load(f; options...)` passes keyword arguments on to the loader.
+""" ->
+function load(s::@compat(Union{AbstractString,IO}), args...; options...)
+    q           = query(s)
+    libraries   = applicable_loaders(q)
+    failures    = Any[]
+    for library in libraries
+        try
+            Library = save_import(library)
+            return Library.load(q, args...; options...)
```

Comment:
It would likely show up for small formatted reads from an `IO`, say a dedicated format with a few bytes for which many can be packed into a single stream. But I agree it's reasonable that the performance shouldn't be a deciding factor in the design for "typical" cases, and it would be great to get past that.

> I'm not sure how the dictionary solution would look like without multiple dispatch...

I guess what bothers me is that we have this awkward hybrid right now. 95% of the time, all the important "dispatch" decisions are being made by dictionary-lookup and module-scoping. The only place dispatch is being used is _within_ a loader/saver module, for those (likely to be few?) cases where a module supports multiple formats. 

For those that do support multiple formats, we can either do the dispatch internally in those modules, or use
```
if sym == :PNG
    # do something
elseif sym == :TIF
    # do something else
end
```
The point being that if 95% of all of our dispatch is going to be via dictionary lookup and module scoping, there seems little point in creating hundreds of new types (all the new `DataFormat{sym}` types) and compiling a brand-new version of all of FileIO's functions for each one.